{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import biom\n",
    "import qiime2\n",
    "import numpy as np\n",
    "import os\n",
    "import skbio\n",
    "\n",
    "import numpy as np\n",
    "from scipy.spatial import ConvexHull\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d.art3d import Poly3DCollection, Line3DCollection\n",
    "\n",
    "\n",
    "from qiime2.plugins.emperor.actions import (plot, biplot)\n",
    "from qiime2.plugins.diversity.actions import (beta_phylogenetic, pcoa, beta)\n",
    "from qiime2.plugins.feature_table.actions import (rarefy, filter_samples)\n",
    "\n",
    "from qiime2.plugins.taxa.actions import collapse\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import rgb2hex\n",
    "\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "# from skbio.stats.ordination import pcoa\n",
    "from scipy import stats\n",
    "%matplotlib inline\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All we will need is a table and associated metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>metab-sample-id</th>\n",
       "      <th>age_units</th>\n",
       "      <th>anonymized_name</th>\n",
       "      <th>cage_food_consumption_g</th>\n",
       "      <th>cage_location</th>\n",
       "      <th>cage_notes</th>\n",
       "      <th>cage_number</th>\n",
       "      <th>collection_timestamp</th>\n",
       "      <th>description</th>\n",
       "      <th>...</th>\n",
       "      <th>physical_specimen_location</th>\n",
       "      <th>qiita_study_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>title</th>\n",
       "      <th>weight_units</th>\n",
       "      <th>age_bin</th>\n",
       "      <th>age_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11829.A45.179.2</td>\n",
       "      <td>A45_179_2</td>\n",
       "      <td>weeks</td>\n",
       "      <td>A45.179.2</td>\n",
       "      <td>44.2</td>\n",
       "      <td>On shelf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A45</td>\n",
       "      <td>4/6/18 11:30</td>\n",
       "      <td>Mouse 179 stool collection 2 of 21</td>\n",
       "      <td>...</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>11829.0</td>\n",
       "      <td>stool</td>\n",
       "      <td>mouse gut metagenome</td>\n",
       "      <td>male</td>\n",
       "      <td>410661.0</td>\n",
       "      <td>10 week IH and IC</td>\n",
       "      <td>g</td>\n",
       "      <td>12.0</td>\n",
       "      <td>10.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11829.A37.150.21</td>\n",
       "      <td>A37_150_21</td>\n",
       "      <td>weeks</td>\n",
       "      <td>A37.150.21</td>\n",
       "      <td>52.7</td>\n",
       "      <td>In chamber</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A37</td>\n",
       "      <td>6/12/18 10:00</td>\n",
       "      <td>Mouse 150 stool collection 21 of 21</td>\n",
       "      <td>...</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>11829.0</td>\n",
       "      <td>stool</td>\n",
       "      <td>mouse gut metagenome</td>\n",
       "      <td>male</td>\n",
       "      <td>410661.0</td>\n",
       "      <td>10 week IH and IC</td>\n",
       "      <td>g</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11829.A35.140.2</td>\n",
       "      <td>A35_140_2</td>\n",
       "      <td>weeks</td>\n",
       "      <td>A35.140.2</td>\n",
       "      <td>37.2</td>\n",
       "      <td>On shelf</td>\n",
       "      <td>Fighting cage</td>\n",
       "      <td>A35</td>\n",
       "      <td>4/6/18 12:00</td>\n",
       "      <td>Mouse 140 stool collection 2 of 21</td>\n",
       "      <td>...</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>11829.0</td>\n",
       "      <td>stool</td>\n",
       "      <td>mouse gut metagenome</td>\n",
       "      <td>male</td>\n",
       "      <td>410661.0</td>\n",
       "      <td>10 week IH and IC</td>\n",
       "      <td>g</td>\n",
       "      <td>12.0</td>\n",
       "      <td>10.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11829.A42.168.11</td>\n",
       "      <td>A42_168_11</td>\n",
       "      <td>weeks</td>\n",
       "      <td>A42.168.11</td>\n",
       "      <td>48.2</td>\n",
       "      <td>In chamber</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A42</td>\n",
       "      <td>5/8/18 11:00</td>\n",
       "      <td>Mouse 168 stool collection 11 of 21</td>\n",
       "      <td>...</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>11829.0</td>\n",
       "      <td>stool</td>\n",
       "      <td>mouse gut metagenome</td>\n",
       "      <td>male</td>\n",
       "      <td>410661.0</td>\n",
       "      <td>10 week IH and IC</td>\n",
       "      <td>g</td>\n",
       "      <td>16.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11829.A41.166.2</td>\n",
       "      <td>A41_166_2</td>\n",
       "      <td>weeks</td>\n",
       "      <td>A41.166.2</td>\n",
       "      <td>41.8</td>\n",
       "      <td>In chamber</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A41</td>\n",
       "      <td>4/6/18 11:00</td>\n",
       "      <td>Mouse 166 stool collection 2 of 21</td>\n",
       "      <td>...</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>11829.0</td>\n",
       "      <td>stool</td>\n",
       "      <td>mouse gut metagenome</td>\n",
       "      <td>male</td>\n",
       "      <td>410661.0</td>\n",
       "      <td>10 week IH and IC</td>\n",
       "      <td>g</td>\n",
       "      <td>12.0</td>\n",
       "      <td>10.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample_name metab-sample-id age_units anonymized_name  \\\n",
       "0   11829.A45.179.2       A45_179_2     weeks       A45.179.2   \n",
       "1  11829.A37.150.21      A37_150_21     weeks      A37.150.21   \n",
       "2   11829.A35.140.2       A35_140_2     weeks       A35.140.2   \n",
       "3  11829.A42.168.11      A42_168_11     weeks      A42.168.11   \n",
       "4   11829.A41.166.2       A41_166_2     weeks       A41.166.2   \n",
       "\n",
       "  cage_food_consumption_g cage_location     cage_notes cage_number  \\\n",
       "0                    44.2      On shelf            NaN         A45   \n",
       "1                    52.7    In chamber            NaN         A37   \n",
       "2                    37.2      On shelf  Fighting cage         A35   \n",
       "3                    48.2    In chamber            NaN         A42   \n",
       "4                    41.8    In chamber            NaN         A41   \n",
       "\n",
       "  collection_timestamp                          description  ...  \\\n",
       "0         4/6/18 11:30   Mouse 179 stool collection 2 of 21  ...   \n",
       "1        6/12/18 10:00  Mouse 150 stool collection 21 of 21  ...   \n",
       "2         4/6/18 12:00   Mouse 140 stool collection 2 of 21  ...   \n",
       "3         5/8/18 11:00  Mouse 168 stool collection 11 of 21  ...   \n",
       "4         4/6/18 11:00   Mouse 166 stool collection 2 of 21  ...   \n",
       "\n",
       "  physical_specimen_location qiita_study_id  sample_type  \\\n",
       "0                     UCSDMI        11829.0        stool   \n",
       "1                     UCSDMI        11829.0        stool   \n",
       "2                     UCSDMI        11829.0        stool   \n",
       "3                     UCSDMI        11829.0        stool   \n",
       "4                     UCSDMI        11829.0        stool   \n",
       "\n",
       "        scientific_name   sex  taxon_id              title weight_units  \\\n",
       "0  mouse gut metagenome  male  410661.0  10 week IH and IC            g   \n",
       "1  mouse gut metagenome  male  410661.0  10 week IH and IC            g   \n",
       "2  mouse gut metagenome  male  410661.0  10 week IH and IC            g   \n",
       "3  mouse gut metagenome  male  410661.0  10 week IH and IC            g   \n",
       "4  mouse gut metagenome  male  410661.0  10 week IH and IC            g   \n",
       "\n",
       "  age_bin age_match  \n",
       "0    12.0      10.5  \n",
       "1    20.0      20.0  \n",
       "2    12.0      10.5  \n",
       "3    16.0      15.0  \n",
       "4    12.0      10.5  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qtab = qiime2.Artifact.load('data/table.qza')\n",
    "meta = pd.read_csv('data/metadata.tsv', sep='\\t')\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Examine how many observations at each timepoint for each condition we have\n",
    "\n",
    "We are interested in how IH (intermittent hypoxia) or IC (hypercapnia) alter the microbiome. Our time variable will be given by \"age_match\" and our group variable will be IH vs. IC vs. Air (control) which in the metadata is called \"exposure_type\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>age_match</th>\n",
       "      <th>exposure_type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">10.0</th>\n",
       "      <th>IC</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Air</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IH</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">10.5</th>\n",
       "      <th>IC</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IH</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Air</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">15.0</th>\n",
       "      <th>IC</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IH</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Air</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">20.0</th>\n",
       "      <th>IC</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IH</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Air</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         count\n",
       "age_match exposure_type       \n",
       "10.0      IC                16\n",
       "          Air               15\n",
       "          IH                15\n",
       "10.5      IC                16\n",
       "          IH                16\n",
       "          Air               15\n",
       "15.0      IC                16\n",
       "          IH                16\n",
       "          Air               15\n",
       "20.0      IC                16\n",
       "          IH                16\n",
       "          Air               13"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(meta[['age_match', 'exposure_type']].value_counts()).sort_values(by='age_match')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualize each time point in 3d\n",
    "Use an appropriate distance metric to generate an ordination. If you are unfamiliar with this step please consult the RPCA tutorial on the gemelli github page, which gives a more detailed explanation of how the RPCA distance metric works and how to use it. Your downstream results will be heavily impacted based on which distance metric you choose to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/rpca-dmat.qza'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# USE YOUR FAVORITE DISTANCE METRIC TO GENERATE DISTANCE MATRIX & ORDINATION\n",
    "from qiime2.plugins.gemelli.actions import rpca\n",
    "result = rpca(qtab)\n",
    "result.biplot.save(f'data/rpca-biplot.qza')\n",
    "result.distance_matrix.save(f'data/rpca-dmat.qza')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# QIIME2 WILL AUTOMATICALLY DO THIS WHEN I REPLACE W PLUGIN\n",
    "tab = qtab.view(biom.Table)\n",
    "ord = result.biplot.view(skbio.OrdinationResults)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-amplicon-2024.5",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
